### PR TITLE
Fix 329 361

### PR DESCRIPTION
**Pull request by kyleflan**
* Fixes issue where using MATCH_BY_COLUMN_NAME led to case sensitivity issues. Replaced with a `quote_identifier` parameter to `write_pandas`, which defaults to False, but if specified as True will quote all identifiers when passing to Snowflake.
* Fixes using `pd_writer` as a method to `DataFrame.to_sql` via `write_pandas` where default columns in a table pre-defined in Snowflake were not having their default values populated

### CODE REVIEW

**Comment by kyleflan** [comment]
@sfc-gh-mkeller do the remaining tests have to be run by your team? Or did I miss a configuration option for the `$PARAMETERS_SECRET` environment variable?

**Comment by sfc-gh-mkeller** [comment]
Hi, @kyleflan 
Thank you for the PR! I'll be more than happy to work with you on this PR.

However; please understand that this is a [behavior change](https://community.snowflake.com/s/article/Pending-Behavior-Change-Log), so I need to get some approvals and announcements before this can be accepted and released.

> @sfc-gh-mkeller do the remaining tests have to be run by your team? Or did I miss a configuration option for the `$PARAMETERS_SECRET` environment variable?

So the `$PARAMETERS_SECRET` is only available to people with write permission to the repository. I need to figure out how to run tests for incoming PRs still, as we just recently moved full development from a private to this public repository.

It is stored in the repository's secrets, so I think after careful inspection of incoming PRs someone with write permission should be able to trigger tests.

If anyone had any input on how to do this properly, I would appreciate the help!

Could you first fix the PR by rebasing onto master? We just introduced the `src` directory.

**Comment by kyleflan** [comment]
Understood about the behavior change. Rebased and pushed.

File: src/snowflake/connector/pandas_tools.py
Comment (sfc-gh-mkeller) [inline src/snowflake/connector/pandas_tools.py:40]: By making this `False` by default is the part that makes this a breaking change. Unfortunately I have still not received approval for this, which is why I have been MIA from your PR @kyleflan

File: test/pandas/test_pandas_tools.py
Comment (sfc-gh-mkeller) [inline test/pandas/test_pandas_tools.py:71]: In general I'd rather have you add one more parametrized argument to the tests than copy paste the whole test just to change one parameter passed down.

File: test/pandas/test_pandas_tools.py
Comment (sfc-gh-mkeller) [inline test/pandas/test_pandas_tools.py:69]: Oof, I don't know what I was thinking when I wrote this...
```suggestion
# Note: since the file will be too small to chunk, this is only testing the put command's syntax
```

**Review by sfc-gh-mkeller — CHANGES_REQUESTED** [review]
@kyleflan I'd suggest breaking this change into 2 parts. One to deal with #361 and another to deal with #329 
Then we could test and merge the #329 even today to unblock you.
If you don't have time to do this please let me know and I can separate things

**Comment by kyleflan** [comment]
@sfc-gh-mkeller how would you like me to separate them? Two new PRs?

**Comment by sfc-gh-mkeller** [comment]
Both that and transforming this into one of them works for me. I’ll let you decide! 🙂

**Comment by kyleflan** [comment]
What if we set `quote_identifiers=True` by default? That will bring it in line with the current behavior, but offer the opportunity for developers to utilize `write_pandas` without quoted identifiers. 

I still think, long term, it makes more sense for the function to operate similar to SQL clients. At that point, it's just a one-line change if we go this route.

**Comment by sfc-gh-mkeller** [comment]
> I still think, long term, it makes more sense for the function to operate similar to SQL clients. At that point, it's just a one-line change if we go this route.

I agree! Yeah, could we flip the default around.
Also please let's merge the 2 tests, you could add a new parametrized parameter to it to have it run with both of the quote_identifiers turned on and off.
I'm sorry for not responding earlier, I'm a bit spread thin nowadays! 😓

**Comment by kyleflan** [comment]
Done. One thing I added to the docstring of `pd_writer` was a description of how to use the `quote_identifiers` arg when using pandas' `DataFrame.to_sql()`. 

I added a test for that: `test_pandas_to_sql()`. However, it requires the user testing it to have the `snowflake-sqlalchemy` package installed, so I check that the user has it installed before running the test.

**Comment by sfc-gh-mkeller** [comment]
Thank you so much @kyleflan .
I like the changes to the code.
But I would like to add some more tests myself to this PR and remove our SQLAlchemy library from the testing for the connector. I can move your test over to our SQLAlchemy repo and then modify your test here not to not test `snowflake-sqlalchemy`.
If I just close this PR don't be alarmed, I'll pull your changes until now into a local branch and then add the test changes there. This procedure will preserve you as the author of these commits, but you should see a mention on this PR anyways.

**Pull request closed without merging**
